noisy-simulator/src/instrument.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This module contains the `Instrument` struct, used to make measurments
+//! in a quantum system.
+
+#[cfg(test)]
+mod tests;
+
+use crate::{operation::Operation, SquareMatrix, TOLERANCE};
+
+/// An instrument is the means by which we make measurements on a quantum system.
+pub struct Instrument {
+    operations: Vec<Operation>,
+    summed_operation: SquareMatrix,
+    summed_effect: SquareMatrix,
+    summed_effect_transpose: SquareMatrix,
+    summed_kraus_operators: Vec<SquareMatrix>,
+}
+
+impl Instrument {
+    /// Creates a new instrument.
+    pub fn new(operations: Vec<Operation>) -> Self {
+        let summed_operation: SquareMatrix = operations.iter().map(|op| op.matrix()).sum();
+        let summed_effect: SquareMatrix = operations.iter().map(|op| op.effect_matrix()).sum();
+        let summed_effect_transpose = summed_effect.transpose();
+        let summed_kraus_operators = summed_kraus_operators(&operations);
+
+        Self {
+            operations,
+            summed_operation,
+            summed_effect,
+            summed_effect_transpose,
+            summed_kraus_operators,
+        }
+    }
+
+    /// Return operation corresponding to the i-th outcome.
+    pub fn operation(&self, i: usize) -> &Operation {
+        &self.operations[i]
+    }
+
+    /// Return the matrix corresponding to the sum over all
+    /// operations in this instrument:
+    /// $$ \sum_i \sum_k K_{ik} \otimes (K_{ik})* $$
+    /// where \otimes denotes the Kronecker product
+    /// and * denotes the complex conjugate (entry-wise).
+    pub fn non_selective_operation_matrix(&self) -> &SquareMatrix {
+        &self.summed_operation
+    }
+
+    /// Return Kraus operators of the operation corresponding to non selective evolution.
+    pub fn non_selective_kraus_operators(&self) -> &Vec<SquareMatrix> {
+        &self.summed_kraus_operators
+    }
+
+    /// Return total effect $(\sum_i \sum_k K_{ik}^\dagger K_{ik})$.
+    pub fn total_effect(&self) -> &SquareMatrix {
+        &self.summed_effect
+    }
+
+    /// Return transposed total effect $(\sum_i \sum_k K_{ik}^\dagger K_{ik})^T$.
+    pub fn total_effect_transposed(&self) -> &SquareMatrix {
+        &self.summed_effect_transpose
+    }
+
+    /// Return number of operations/outcomes in this instrument.
+    pub fn num_operations(&self) -> usize {
+        self.operations.len()
+    }
+}
+
+fn summed_kraus_operators(operations: &[Operation]) -> Vec<SquareMatrix> {
+    let choi_matrix: SquareMatrix = operations
+        .iter()
+        .map(|op| {
+            op.kraus_operators()
+                .iter()
+                .map(|m| {
+                    // choi_matrix += vec(K) * vec(K)^dagger
```

Comment:
Is this left over code or a comment of what the code is doing?

Reply:
A comment of what the code is doing

Reply:
Make the comment more clear